src/expr/src/scalar/func.rs

```diff
@@ -6932,6 +6936,101 @@ fn make_mz_acl_item<'a>(datums: &[Datum<'a>]) -> Result<Datum<'a>, EvalError> {
     }))
 }
 
+fn array_fill<'a>(
+    datums: &[Datum<'a>],
+    temp_storage: &'a RowArena,
+) -> Result<Datum<'a>, EvalError> {
+    const MAX_SIZE: usize = 1 << 28 - 1;
+    const NULL_ARR_ERR: &str = "dimension array or low bound array";
+    const NULL_ELEM_ERR: &str = "dimension values";
+
+    let fill = datums[0];
+    if matches!(fill, Datum::Array(_)) {
+        return Err(EvalError::Unsupported {
+            feature: "array_fill with arrays".to_string(),
+            issue_no: None,
+        });
+    }
+
+    let arr = match datums[1] {
+        Datum::Null => return Err(EvalError::MustNotBeNull(NULL_ARR_ERR.to_string())),
+        o => o.unwrap_array(),
+    };
+
+    let dimensions = arr
+        .elements()
+        .iter()
+        .map(|d| match d {
+            Datum::Null => Err(EvalError::MustNotBeNull(NULL_ELEM_ERR.to_string())),
+            d => Ok(usize::cast_from(u32::reinterpret_cast(d.unwrap_int32()))),
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let lower_bounds = match datums.get(2) {
+        Some(d) => {
+            let arr = match d {
+                Datum::Null => return Err(EvalError::MustNotBeNull(NULL_ARR_ERR.to_string())),
+                o => o.unwrap_array(),
+            };
+
+            arr.elements()
+                .iter()
+                .map(|l| match l {
+                    Datum::Null => Err(EvalError::MustNotBeNull(NULL_ELEM_ERR.to_string())),
+                    l => Ok(isize::cast_from(l.unwrap_int32())),
+                })
+                .collect::<Result<Vec<_>, _>>()?
+        }
+        None => {
+            vec![1isize; dimensions.len()]
+        }
+    };
+
+    if lower_bounds.len() != dimensions.len() {
+        return Err(EvalError::ArrayFillWrongArraySubscripts);
+    }
+
+    let fill_count: usize = dimensions
+        .iter()
+        .cloned()
+        .map(Some)
+        .reduce(|a, b| match (a, b) {
+            (Some(a), Some(b)) => a.checked_mul(b),
+            _ => None,
```

Comment:
When do we expect this case? Seems untested.

Reply:
This test case triggers it:

```
SELECT array_fill(1, ARRAY[4, -199, 3]);
```

This is essentially 4 * (usize::MAX - 199) * 3, the first two operations will overflow the `usize`, which means that the input to the reduce will become `(None, Some(3)`, which triggers the last line in this comment.

Reply:
I added a `panic!` there, but am still not seeing it:
```
materialize=> SELECT array_fill(1, ARRAY[4, -199, 3]);
ERROR:  array size exceeds the maximum allowed (134217728 bytes)
```

Reply:
Sorry about that--they're u32s getting multiplied, which takes longer to overflow usize; added a test case which I verified hits the last match branch but this didn't change the behavior at all:

```
SELECT array_fill(8, ARRAY[-1, -1, -1, -1]);
```